test/rnn_seq_api.hpp

```diff
@@ -0,0 +1,1663 @@
+                                                 hidVec),
+                    nodhx,
+                    nodcx};
+        }
+        else
+            return {std::move(packed_dInput),
+                    std::move(d_hidden_state),
+                    std::move(d_cell_state),
+                    nodhx,
+                    nodcx};
+    }
+
+    WRWResObj wrw(const miopen::SeqTensorDescriptor& xDesc,
+                  const miopen::SeqTensorDescriptor&,
+                  const std::vector<T>& xData,
+                  const std::vector<T>& hxData,
+                  const std::vector<T>& doutData,
+                  std::vector<T>& reserveSpace,
+                  std::vector<T>& workSpace,
+                  bool nohx) const override
+    {
+        assert(checkSeqTensor(xDesc));
+
+        auto batch_seq = xDesc.GetBatchesPerSequence();
+
+        std::vector<int> batch_seq_downgrade(batch_seq.cbegin(), batch_seq.cend());
+
+        size_t seq_len    = batch_seq.size();
+        size_t batch_size = batch_seq.at(0);
+
+        bool is_state_tensor_zip_req = (batch_size != xDesc.GetLengths()[0]);
+        bool is_hx_zip_req           = is_state_tensor_zip_req && !nohx;
+
+        std::vector<T> dwei_data(UniRNNCPUWeightSize(
+            rnnMode, rnnDesc.nLayers, hidVec, inVec, biasMode, inputMode, dirMode));
+
+        UniformRNNBwdWeightCPUVerify(
+            useDropout,
+            xData,
+            dwei_data, // (output) [ input_state_weight_trans
+                       // hidden_state_weight0_trans
+                       // input1_trans hidden1_trans ...
+                       // output_weight; bidirectional
+                       // reversed weights ]
+            is_hx_zip_req ? zipStateVectorTensor(
+                                hxData, hiddenLayers, xDesc.GetLengths()[0], batch_size, hidVec)
+                          : hxData, // initial hidden state
+            doutData,
+            batch_seq_downgrade, // input batch size
+            inVec,               // input data length
+            seq_len,             // Number of iterations to unroll over
+            dirMode,             // whether using bidirectional net
+            biasMode,            // whether using bias
+            hiddenLayers,        // 1 by numlayer (number of stacks of hidden
+                                 // layers) for unidirection, 2 by numlayer for
+                                 // bidirection
+            batch_size,          // equal to input batch size in_n[0]
+            hidVec,              // hidden state number
+            outVec,              // 1 by hy_h related function for unidirection, 2
+                                 // by hy_h related function for bidirection
+            rnnMode,
+            inputMode,
+            reserveSpace,
+            workSpace,
+            nohx);
+
+        return {std::move(dwei_data)};
+    }
+
+private:
+    bool checkSeqTensor(const miopen::SeqTensorDescriptor& desc) const
```

Comment:
[note] This function is used in debug builds only. I recommend running clang-tidy locally.

Reply:
[resolved] AFAICS CI has passed.